### ISSUE TITLE
StringIndexOutOfBoundsException

### ISSUE THREAD

**Issue by tdzedaj** [reactor/reactor-core#2547]
reactor.core.Scannable, method stepName

This code:

```
name = name
				.substring(0, name.indexOf('$'))
				.substring(name.lastIndexOf('.') + 1);
```
throws .StringIndexOutOfBoundsException for name: 

`ScopePassingSpanSubscriber{subscriber=reactor.core.publisher.MonoNext$NextSubscriber@1bf6ebc, parent=org.springframework.cloud.sleuth.api.noop.NoOpTraceContext@7110297e}`

I don't think I need to explain why

**Comment by simonbasle** [comment]
This is more of an issue of Sleuth's `ScopePassingSubscriber` not overriding `stepName` when it overrides `toString`/`name`. That said, we could catch exceptions and fallback to `name()` in `stepName()` maybe?

**Comment by simonbasle** [comment]
cc @marcingrzejszczak

**Comment by marcingrzejszczak** [comment]
@tdzedaj can you create an issue in spring-cloud-sleuth? @simonbasle should we just remove `toString()` ?

**Comment by simonbasle** [comment]
@marcingrzejszczak I guess it can make sense to have a customized `toString()`, and here the best course of action is to override `stepName()` to return something like `subscriber.stepName() + "(traced)"` or something similar (ie rely on the `stepName()` of the wrapped subscriber).

**Comment by tdzedaj** [comment]
Regardless of what is in spring-cloud-sleuth, in this particular method there is also an bug. The assumption is probably made that toString () will return a properly formatted string containing a dot before the dollar sign.

In this case, it is not checked whether a dot is present and is before a dollar sign.

**Comment by simonbasle** [comment]
superseded by #2526 and fixed by 4de5ab08aa1efa0496de4d2f75adb58f4308c935
